torch/distributed/traceable_collectives.py

```diff
@@ -0,0 +1,188 @@
+  compiler: trace these ops with plain-old-data schemas, then choose how to lower them.
+  eager: execute these 'functional' ops which in eager return AsyncCollectiveTensor subclasses,
+         automatically calling .wait() on underlying/hidden async 'work' obj only when fed to
+         a downstream op.
+
+Issues:
+* Where should these ops live? Couldn't `import torch` if putting these ops in existing torch.distributed files
+* How can we make these ops work in eager without manually enabling python_dispatacher mode?
+
+"""
+
+"""
+FIXME for this to work correctly we need to change Work to internally hold no reference to the tensor.
+"""
+
+#FIXME we do this way cuz we can't use tensor __eq__, we want id()
+#TODO use a weakref callback
+class IdKeyWeakDict:
+    def __init__(self):
+        self.kvs = []
+
+    def add(self, key, val):
+        self.kvs.append((weakref.ref(key), val))
+
+    def get_and_remove(self, key):
+        new_arr = []
+        val = None
+        for k, v in self.kvs:
+            this_k = k()
+            if this_k == None:
+                continue
+            if id(this_k) == id(key):
+                val = v
+            else:
+                new_arr.append((k, v))
+        self.kvs = new_arr
+        return val
+
+tensor_to_work = IdKeyWeakDict()
+
+lib = torch.library.Library("tr_c10d", "DEF")
+lib.define("wait(Tensor self) -> Tensor")
+
+impl_lib = torch.library.Library("tr_c10d", "IMPL",  "CPU")
+
+def _wait_tensor(tensor: torch.Tensor):
+    w = tensor_to_work.get_and_remove(tensor)
+    if w:
+        w.wait()
+    return tensor
+
+impl_lib.impl("wait", _wait_tensor)
+
+def wait_tensor(tensor):
+    return torch._ops.ops.tr_c10d.wait(tensor)
+
+class AsyncCollectiveTensor(torch.Tensor):
+    r"""
+    A Tensor subclass that is only used in eager mode, to hold a 'work' object
+    and then wait on it before invoking a real op.
+
+    Usage, from inside functional collective:
+    def functional_collective(input):
+        input = input.clone()
+        mutated_input, work = c10d.{inplace_collective}(input)
+        return AsyncCollectiveTensor(mutated_input, work)
+    """
+    _tensor: torch.Tensor
+
+    # disable __torch_function__ so that CommTensor can recursively dispatch
+    # with ProxyTorchDispatchMode in make_fx
+    __torch_function__ = _disabled_torch_function_impl
+
+    @staticmethod
+    def __new__(cls, tensor: torch.Tensor):
+        t = tensor
+        r = torch.Tensor._make_subclass(cls, t, require_grad=t.requires_grad)
+        r._tensor = tensor  # type: ignore[attr-defined]
+        return r
+
+    def __repr__(self):
+        return f"AsyncCollectiveTensor({self._tensor})"
+
+    @classmethod
+    def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
+        def unwrap(e: Any):
+            if isinstance(e, AsyncCollectiveTensor):
+                return wait_tensor(e._tensor)
+            return e
+
+        # TODO what happens when func is an inplace? (add_) -
```

Comment:
im still not sure what we'll do about this.

is your idea that it's OK to wait_tensor multiple times?  (both functionally and not slow)

Reply:
Yes, I think wait_tensor should be idempotent, to simplify usage.

Current implementation makes the strong assumptions like there's a single compute stream. We can harden against that.